### PR TITLE
Adds support of broadcast and reduction with shift

### PR DESCRIPTION
**Pull request by naoyam**
Follow-up to #809. Adds support of broadcast and reduction.

Additionally, refactored how `kir::Predicate` is created. For exprs requiring block sync, the predicate is created at the indexing step, but that doesn't need to be, so it's done at the same time as other exprs. This also makes `PredicateType::InternalSync` unnecessary, so removed it.

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/lower_shift.cpp [hunk 1]
```diff
@@ -87,6 +82,19 @@ void ShiftPredicateInserter::insert(
   //   }
   // }
 
+  // If the expr involves a thread-block barrier, set the predicate of
+  // the expre with shift_pred. No padding should be necessary.
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/lower_shift.cpp]: note to self: doublecheck if this is really true

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_shift.cpp]: It should be safe to omit padding. In fact, it's probably safe to omit for all non-shift exprs.

**Review by rdspring1 — COMMENTED** [review]
The predicate changes LGTM

File: torch/csrc/jit/codegen/cuda/lower_allocation.cpp [hunk 2]
```diff
@@ -127,7 +127,21 @@ class AllocationInserter : public kir::MutableIrVisitor {
          init_loop_it != init_dims.rend();
          ++init_loop_it) {
       auto id = *init_loop_it;
-      kir::ForLoop* new_loop = ir_builder.create<kir::ForLoop>(id);
+      kir::ForLoop* new_loop = nullptr;
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/lower_allocation.cpp:130]: How did this work previously?

**Review by csarofeen — APPROVED** [review]
Seems fine to me.

File: torch/csrc/jit/codegen/cuda/lower_allocation.cpp (hunk 2 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/lower_allocation.cpp:130]: It did not work previously. There was no test exercising this case, i.e., reduction tensors with halo.

**Pull request merged**